twister: introduce twister-level test timeout management

evgeniy-paltsev opened the pull request:
Twister allows us to control maximum execution time for each test with timeout value in test's `.yaml` configuration and platform level timeout multiplier which allows us to tweak timeout value for specific platform.

However, sometimes we want to additionally adjust tests timeouts when running twister. This is especially useful in case of simulation platform as simulation time may depend on the host speed & load, we may select different simulation method (i.e. cycle accurate but slower one), etc...

Let's introduce global (twister-level) timeout multiplier option to handle these cases.

evgeniy-paltsev committed:
twister: introduce twister-level timeout multiplier

Twister allows us to control maximum execution time for each test
with timeout value in test's .yaml configuration and
platform level timeout multiplier which allows us to tweak
timeout value for specific platform.

However, sometimes we want to additionally adjust tests timeouts
when running twister. This is especially useful in case of
simulation platform as simulation time may depend on the host
speed & load, we may select different (i.e. cycle accurate but
slower one) simulation method, etc...

Let's introduce global (twister-level) timeout multiplier option.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

abrodkin commented:
Would be good to provide some rationale behind that change.

evgeniy-paltsev committed:
twister: tests: handler: fix incorrect timeout mock

There is clearly an issue with current timeout value
timeout into mocked_instance as this value is used only
once before this mock happen, so this mock is actually useless
and we always use default value. Fix that.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

hakehuang reviewed (dismissed):
one more thing, please update the twister document. others looks good to me

evgeniy-paltsev commented:
Hi @hakehuang,

> one more thing, please update the twister document. others looks good to me

But do we document twster options in some Zephyr documentation file? I didn't find any and from what I see we only document options in `help` section of argument parser.

In [Test Runner](https://docs.zephyrproject.org/latest/develop/test/twister.html) docs we document the `.yaml` test bindings (which are used by twister) and only mention few of twister options when describing the related functionality.

I've also checked few recent PRs where we were adding new twister args -  and we do not provide any documentation in addition to what we have in arg parser help section:
https://github.com/zephyrproject-rtos/zephyr/pull/57806/files
https://github.com/zephyrproject-rtos/zephyr/pull/52893/files
https://github.com/zephyrproject-rtos/zephyr/pull/51441/files

nashif commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -196,6 +196,11 @@ def add_parse_arguments(parser = None):
         help="""Only run device tests with current artifacts, do not build
              the code""")
 
+    test_or_build.add_argument("--timeout-multiplier", type=float, default=1,
```
why is this being added to test_or_build? 
What if I want to use this argument with --test-only?, right now I am getting:

```twister: error: argument --test-only: not allowed with argument --timeout-multiplier```

evgeniy-paltsev replied:
Yep, indeed it's not correct. Was confused by a bit misleading `test_or_build name`. :)

Now fixed.

nashif commented on scripts/pylib/twister/twisterlib/handlers.py:
```diff
@@ -94,6 +93,11 @@ def __init__(self, instance, type_str="build"):
 
+    def test_timeout(self):
+        return math.ceil(self.instance.testsuite.timeout *
+                         self.instance.platform.timeout_multiplier *
```
Isnt this going to "exponentially" increase the timeout multiplying this again by whatever the platform has?

So if I have timeout of 100 and platform has a 4 multiplier and I use the new argument with 4, it will be 100 * 4 * 4.

I think we need to ignore what platform has provided if multiplier is provided on command line.

ruuddw replied:
Not sure, then running a simulation of different platforms on a 'slow' host would have the same platform penalty for all platforms (e.g. single and multi-core ISS). They feel like different things to me, better apply them both.

nashif replied:
> This is especially useful in case of simulation platform as simulation time may depend on the host speed & load, we may select different simulation method (i.e. cycle accurate but slower one), etc...

`self.instance.platform.timeout_multiplier` was added to address this issue on a platform level, now we are adding an option that is used on the global level, they are mutually exclusive IMO.

ruuddw replied:
For 'real' platforms indeed that works. The issue addressed here, however, is different, mostly for simulation platforms. It is not about the relative speed/performance of a platform (e.g. fast or slow simulator), but about the speed of the machine running the simulator. Current platforms have been calibrated for Zephyr CI, but on other environments may need 'tuning'. That's about the development host, not about the specific platform being simulated.

nashif replied:
ok, my point here is that you use this option only with a set of platforms, not when you call twister globally, i.e. 

`twister -p nsim_hs5x_smp_12cores --timeout-multiplier 10`

but looking in the tree:

```
identifier: nsim_hs5x_smp_12cores
name: Multi-core HS5x nSIM simulator
type: sim
simulation: mdb-nsim
simulation_exec: mdb
arch: arc
toolchain:
  - zephyr
  - arcmwdt
  - cross-compile
supported:
  - smp
testing:
  timeout_multiplier: 4
  ignore_tags:
    - net
    - bluetooth
```

There is already a multiplier, so the timeout will be multiplied again by 4 (effective multiplier will be 40). Why not call twister with the value of 40 instead and avoid having to do special calculations?

Anyways, it is time you will spend waiting for something to timeout unnecessarily IMO.

ruuddw replied:
right, this example is about a 12-core SMP simulation target. We run that together with a 1-core ARC simulation in our internal regression testing environment, or maybe on a local developer PC. the x4 for the platform handles the 12 vs 1 core simulation speed. The global parameter handles the difference between running nSIM on local PC, Synopsys compute farm, or Zephyr CI.

evgeniy-paltsev replied:
@nashif 

> I think we need to ignore what platform has provided if multiplier is provided on command line.

It's intentionally implemented this way otherwise it would be really counterintuitive for user. I.e. we may want to increase timeout, but adding `--timeout-multiplier` may decrease timeout  (if the value in board.yaml is bigger than the value we've selected via `--timeout-multiplier`). It's even worse if several platforms are selected as we may increase timeouts for one and decrease timeouts for another.

So yes, we should use it as a coeficient rather than overrde platform provided value.

nashif replied:
> It's even worse if several platforms are selected as we may increase timeouts for one and decrease timeouts for another.

Do not understand how is the total timeout is  going to decrease, can you explain?

nashif replied:
then, in this case, do:

`twister -p board_slow -p board_fast --timeout-multiplier 2`, i.e. use the the multiplier of the slowest.

nashif replied:
> PS: I don't see any reason to override `timeout_multiplier` value in board configuration with twister option. They serve diferent purpose - one for tweaking timeout because of platform specifics (i.e. slow HW / simulator) another one for tweaking timeout because of host / launch specifics (significant host load due to overcommit, extra delay for simulator / debugger licence check, different simulation method usage / etc...)

The timeout will mostly be impacted by the simulator and less by the host being slow. If the host is slow, it does not mean  all tests will run slower.

nashif commented on scripts/pylib/twister/twisterlib/handlers.py:
```diff
@@ -94,6 +93,11 @@ def __init__(self, instance, type_str="build"):
         self.args = []
         self.terminated = False
 
+    def test_timeout(self):
```
I would call it `get_test_timeout`, `test_` pattern applies to rely tests usually, for readability.

evgeniy-paltsev replied:
Fair enough. I've changed it to `get_test_timeout`

nashif commented on scripts/tests/twister/test_handlers.py:
```diff
@@ -1900,14 +1903,13 @@ def mock_cputime(pid):
-    type(mocked_instance.testsuite).timeout = mock.PropertyMock(
-        return_value=timeout
-    )
+    handler.options = mock.Mock(timeout_multiplier=1)
```
@LukaszMrugala pleasse review

hakehuang commented:
> Hi @hakehuang,
> 
> > one more thing, please update the twister document. others looks good to me
> 
> But do we document twster options in some Zephyr documentation file? I didn't find any and from what I see we only document options in `help` section of argument parser.
> 
> In [Test Runner](https://docs.zephyrproject.org/latest/develop/test/twister.html) docs we document the `.yaml` test bindings (which are used by twister) and only mention few of twister options when describing the related functionality.
> 
> I've also checked few recent PRs where we were adding new twister args - and we do not provide any documentation in addition to what we have in arg parser help section: https://github.com/zephyrproject-rtos/zephyr/pull/57806/files https://github.com/zephyrproject-rtos/zephyr/pull/52893/files https://github.com/zephyrproject-rtos/zephyr/pull/51441/files

@teburd, many twister options are not documented as they are existing before this document starting, and for any new options  that can be helpful to user, I would recommend to add documents in [Test Runner](https://docs.zephyrproject.org/latest/develop/test/twister.html), for this one, I see this is for customization, and it would be hard to understand how to use, if you just leave simple lines.

evgeniy-paltsev committed:
doc: twister: add chapter about managing tests timeouts

Document timeout_multiplier board.yaml option and
--timeout-multiplier twister option.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev commented:
Hi @hakehuang,

I've added documentation, please take a look.

nashif reviewed (approved):
will revisit this later, as my comments only apply if board have multiplier already, and we do not have many of those.
